fix(terminal): report terminal window size in pixels

lumynou5 opened the pull request:
Window size in pixels is required by some programs, reporting it can help those programs work inside NeoVim terminals.

Closes #8259.

However, this patch doesn't make [`icat`](https://sw.kovidgoyal.net/kitty/kittens/icat/) work. Instead, it just makes the program show another error, because the problem is not only related to window sizes in pixels but how escape sequences (specifically, APC) are handled.  See #4349, #20176.

lumynou5 committed:
fix(terminal): report terminal window size in pixels

Window size in pixels is required by some programs, reporting it can
help those programs work inside NeoVim terminals.

See also: #8259, #12991

justinmk commented on src/nvim/os/pty_process_unix.c:
```diff
@@ -171,7 +171,13 @@ int pty_process_spawn(PtyProcess *ptyproc)
+  ptyproc->winsize = (struct winsize){
+    ptyproc->height,
+    ptyproc->width,
+    ptyproc->winsize.ws_xpixel / ptyproc->winsize.ws_col * ptyproc->width,
+    ptyproc->winsize.ws_ypixel / ptyproc->winsize.ws_row * ptyproc->height,
```
use parens to make the order of operations obvious
```suggestion
    (ptyproc->winsize.ws_xpixel / ptyproc->winsize.ws_col) * ptyproc->width,
    (ptyproc->winsize.ws_ypixel / ptyproc->winsize.ws_row) * ptyproc->height,
```

justinmk reviewed (approved):
Nice, thanks. If no one else has comments, ping me if this isn't merged in a couple days

justinmk commented:
I wonder if this fixes any of the log messages that were showing in Windows CI https://github.com/neovim/neovim/blob/cf9f002f31c8b4d9d42912a3f45f5d3db4462fd9/test/testutil.lua#L95

lumynou5 committed:
refactor: add parentheses for readability

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

lumynou5 commented:
Currently the test is failed.  When stdout is redirected, `ioctl` calls fail with errno 25, and the program crashes due to invalid division.
Adding a check to set pixel sizes to `0` in case of failed calls can help, but it seems not a desirable solution?

```c
  int res = ioctl(1, TIOCGWINSZ, &ptyproc->winsize);
  ptyproc->winsize = (struct winsize){
    ptyproc->height,
    ptyproc->width,
    res ? 0 : (ptyproc->winsize.ws_xpixel / ptyproc->winsize.ws_col) * ptyproc->width,
    res ? 0 : (ptyproc->winsize.ws_ypixel / ptyproc->winsize.ws_row) * ptyproc->height,
  };
```

lumynou5 committed:
fix(terminal): report winszie in pixels as 0 if it's not a TTY

Window sizes in pixels now have a fallback value 0 when NeoVim is not
running in a TTY.  This prevents NeoVim from crashing if it's running in
an IDE's integrated terminal, a terminal multiplexer, or a testing, etc.

zeertzjq commented:
It seems this doesn't actually solve anything, and may not be correct. For example, when running Nvim with `--headless` and attaching another TUI with `--remote-ui`, this will use the pixel size of the terminal running the `--headless` instance instead of the one running the `--remote-ui` instance.

lumynou5 commented:
> It seems this doesn't actually solve anything, and may be incorrect. For example, when running Nvim with `--headless` and attaching another TUI with `--remote-ui`, this will use the pixel size of the terminal running the `--headless` instance instead of the one running the `--remote-ui` instance.

Thanks. I didn't think about remote. I'll try to solve it.
